src/main/scala/inox/solvers/SolverResponses.scala

```diff
@@ -99,11 +99,14 @@ object SolverResponses {
     }).asInstanceOf[Response[M,C]]
 
     def convert[M1,T1,M2,T2](resp: Response[M1,Set[T1]], cm: M1 => M2, cc: Set[T1] => Set[T2]): Response[M2,Set[T2]] =
+      convertMaybe(resp, (m1: M1) => Some(cm(m1)), cc)
+
+    def convertMaybe[M1,T1,M2,T2](resp: Response[M1,Set[T1]], cm: M1 => Option[M2], cc: Set[T1] => Set[T2]): Response[M2,Set[T2]] =
```

Comment:
It seems to me like there's some confusion with where the `tryZ3` logic is invoked. Some calls to `model.eval` will be wrapped in a try-catch while others won't be so it's unclear whether we'll end up with `Unknown` or a partial model after the call to `convertMaybe`.

I think it would make more sense to keep only the `convert` call and make sure to wrap all calls into the `Z3Model` class with `tryZ3` in `Z3Native.extractModel`. We can then adapt the `UnrollingSolver` to always check if search was interrupted _after_ having converted the model and return `Unknown` in that case (since the model will be partial): https://github.com/epfl-lara/inox/blob/2f402f9bf9541cc061558e525fdd53efa5f86a87/src/main/scala/inox/solvers/unrolling/UnrollingSolver.scala#L589-L590

Reply:
I like that approach; scattering `tryZ3` a bit everywhere is indeed not very clean :p

Reply:
I think you're mapping to `Unknown` too early. We need to make sure that the model extraction performed by `CheckResult.cast` never results in a partial model. In your implementation, if an interruption occurred between the solver result and extraction, we could still end up with an incomplete model.

I believe the logic should look something like:
```scala
def cast(...) = {
  val res = config.convert(config.cast(resp), extractSimpleModel, decodeAssumptions)
  if (interrupted) Unknown else res
}
```

Reply:
Good catch, thanks!

Reply:
Hopefully the pushed change correctly addresses that point!

Reply:
Looking good!

Sorry I realized I missed a similar check in the `Validate(model)` case: https://github.com/epfl-lara/inox/blob/822692640f458a81f48de79975269fbca261dc61/src/main/scala/inox/solvers/unrolling/UnrollingSolver.scala#L690

Maybe it suffices to move the check into the `CheckResult.apply` method (and only if `checkModels` is false the the model will actually be evaluated otherwise), something like:
```scala
object CheckResult {
  def cast(resp: SolverResponse[underlying.Model, Set[underlying.Trees]]): CheckResult =
    CheckResult(config.convert(config.cast(resp), extractSimpleModel, decodeAssumptions))

  def apply[M <: Model, A <: Assumptions](resp: config.Response[M, A]) = resp match {
    case SatWithModel(model) if !checkModels && abort => new CheckResult(config cast Unknown)
    case _ => new CheckResult(resp)
  }

  def unapply(res: CheckResult): Option[config.Response[Model, Assumptions]] = Some(res.response)
}
```

Note that `abort` is basically the same as `context.interruptManager.isInterrupted`.

Reply:
Sometimes the solvers return a model despite/when they get interrupted, so we can still use these when they appear if we run `checkModels` since the check ensures the model is indeed valid.

Reply:
Good to know, thanks for the explanation :)